### ISSUE TITLE
(Feature request) Generalized Trilium tree

### ISSUE THREAD

**Issue by agentydragon** [zadam/trilium#3263]
### Describe feature

# TL;DR

Currently Trilium tree can be thought of as a DAG which you manage in the tree UI. This DAG is defined by the relation "X is in the tree immediately under Y", and this relation is transitive. (Or, rather, the "X is anywhere in the tree under Y" is that relation's transitive closure.)

My brain does not work by having a single DAG. My brain has multiple incompatible DAGs which arrange the same nodes with different relations.

I propose to generalize the tree feature by:

* Allowing the user to declare that a particular relation in their database is transitive
* Generalizing the Trilium tree view to display *any* relation that is transitive in the user's database.

# Description

for my database the “main Trilium tree" structure is not really useful. I represent different sorts of relations by putting notes under other notes - sometimes “this is a subtopic”, sometimes “this is a resource for this thing”, sometimes “I did this thing on this day”.

the tree child-parent structure is mostly interpreted by Trilium as transitive. if X is parent of Y who is parent of Z, the tree sees it as it also “transitively implying” X being ancestor of Z, but that doesn't make sense if X → Y and Y → Z are different types of relationships.

I think I would honestly benefit from Trilium not having a tree at all.

It is useful for me to see several tree views of my graph-structured database - some of my relations are in fact transitive, like the relation between years, months, days and things I worked on that day. That's basically how my “journal” note is now structured.

But because Trilium has only one privileged tree structure, that structure for me mixes up several different trees into a huge mess that is not even a transitive relation.

I'd ideally like a thing where I'd declare “the relation ‘isAncestorOf’ or ‘thingHappenedDuringTime’ in my database is transitive”. then Trilium would create one tree view for each of those transitive relations. because Trilium allows cloning, Trilium's "tree' is actually technically speaking a rooted DAG. and ideally current tree branching UI would let me manage the tree implied by the relation. just like right now, there would probably be a difference between “R(X,Y) holds because user explicitly added it” and “R(X,Y) can be inferred from the fact that R(X,Z) and R(Z,Y)” - analogous to how tree currently has direct and indirect ancestors.

**Comment by agentydragon** [comment]
related: the "share this note" switch, as far as I know, works by sharing the whole subtree under the note - you can't say "share this note but not its children". this also doesn't make sense for my tree structure.

**Comment by zadam** [comment]
Hi, I get what you're saying. I also sometimes wonder what kind of metamodel would be the best for modeling knowledge and have already pretty much dismissed the tree. The trouble is that you often need to use it at scale to evaluate how good/bad it is (many things work well on a small scale).

In the case of Trilium, the tree is there to stay - it's so central to pretty much everything in the application, that taking it out (or generalizing it based on your suggestion) would require a complete rewrite and would probably turn it into a different application.

Most of the foundational design choices like this were made back in 2017, and I have doubts about some of them. My philosophy is to stick with them for the sake of stability and see where they lead us.

**Comment by agentydragon** [comment]
Sure. Feel free to keep on the backburner for "far-future design ideas".

**Comment by eliandoran** [comment]
I agree with @zadam here, at this point in time it would be impossible to change the entire model, as it would transform Trilium into something else and I'm pretty sure a great chunk of people would resist a complete change in their note structure.
